pkg/apis/servicecatalog/plan_reference.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicecatalog
+
+import (
+	"fmt"
+	"strings"
+)
+
+// ClassSpecified checks that at least one class field is set.
+func (pr PlanReference) ClassSpecified() bool {
+	return pr.ClusterServiceClassExternalName != "" ||
+		pr.ClusterServiceClassExternalID != "" ||
+		pr.ClusterServiceClassName != ""
+}
+
+// PlanSpecified checks that at least one plan field is set.
+func (pr PlanReference) PlanSpecified() bool {
+	return pr.ClusterServicePlanExternalName != "" ||
+		pr.ClusterServicePlanExternalID != "" ||
+		pr.ClusterServicePlanName != ""
+}
+
+// GetSpecifiedClass returns the user-specified class value from either:
+// * ClusterServiceClassExternalName
+// * ClusterServiceClassExternalID
+// * ClusterServiceClassName
+func (pr PlanReference) GetSpecifiedClass() string {
+	if pr.ClusterServiceClassExternalName != "" {
+		return pr.ClusterServiceClassExternalName
+	}
+
+	if pr.ClusterServiceClassExternalID != "" {
+		return pr.ClusterServiceClassExternalID
+	}
+
+	if pr.ClusterServiceClassName != "" {
+		return pr.ClusterServiceClassName
+	}
+
+	return ""
+}
+
+// GetSpecifiedPlan returns the user-specified plan value from either:
+// * ClusterServicePlanExternalName
+// * ClusterServicePlanExternalID
+// * ClusterServicePlanName
+func (pr PlanReference) GetSpecifiedPlan() string {
+	if pr.ClusterServicePlanExternalName != "" {
+		return pr.ClusterServicePlanExternalName
+	}
+
+	if pr.ClusterServicePlanExternalID != "" {
+		return pr.ClusterServicePlanExternalID
+	}
+
+	if pr.ClusterServicePlanName != "" {
+		return pr.ClusterServicePlanName
+	}
+
+	return ""
+}
+
+// GetClassFilterFieldName returns the appropriate field name for filtering
+// a list of service catalog classes by the PlanReference.
+func (pr PlanReference) GetClassFilterFieldName() string {
+	if pr.ClusterServiceClassExternalName != "" {
+		return "spec.externalName"
+	}
+
+	if pr.ClusterServiceClassExternalID != "" {
+		return "spec.externalID"
+	}
+
+	return ""
+}
+
+// GetPlanFilterFieldName returns the appropriate field name for filtering
+// a list of service catalog plans by the PlanReference.
+func (pr PlanReference) GetPlanFilterFieldName() string {
```

Comment:
shouldn't this also return `metadata.name` ?

Reply:
`metadata.name` isn't one of the supported fields for filtering during a list. Can we change that? It would be great to not have to use a Get sometimes and a filtered List other times.

Reply:
from a user perspective, it doesn't make that much sense to filter by name because those are UUIDs, no?

Reply:
Yeah the name is a uuid. If we added filtering by metadata.name, it would be purely for our own convenience in the controller code.

Reply:
I ran into a situation where I wanted to filter by name as a uuid just the other day

Reply:
Ok, cool 👍